tests/src/test/scala/org/apache/openwhisk/core/containerpool/test/ContainerPoolTests.scala

```diff
@@ -753,14 +758,211 @@ class ContainerPoolTests
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
+
+    containers(0).expectMsg(Remove)
+    containers(1).expectMsg(Remove)
+    containers(0).send(pool, ContainerRemoved(false))
+    containers(1).send(pool, ContainerRemoved(false))
+
+    // Because already supplemented the prewarmed container, so currentCount should equal with initialCount
+    stream.toString should include(s"currentCount: ${initialCount} prewarmed container")
+
+    // the desiredCount should equal with minCount because cold start didn't happen
+    stream.toString should include(s"needs ${minCount} desired prewarmed container")
+    // Previously created prewarmed containers should be removed
+    stream.toString should include(s"removed ${initialCount} expired prewarmed container")
+
+    stream.reset()
+    val action = ExecutableWhiskAction(
+      EntityPath("actionSpace"),
+      EntityName("actionName"),
+      exec,
+      limits = ActionLimits(memory = MemoryLimit(memoryLimit)))
+    val run = createRunMessage(action, invocationNamespace)
+    // 2 code start happened
+    pool ! run
+    pool ! run
+    containers(2).expectMsg(run)
+    containers(3).expectMsg(run)
+
+    pool ! AdjustPrewarmedContainer
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
+
+    // Because already removed expired prewarmed containrs, so currentCount should equal with 0
+    stream.toString should include(s"currentCount: 0 prewarmed container")
+    // the desiredCount should equal with 2 due to cold start happened
+    stream.toString should include(s"needs 2 desired prewarmed container")
+
+    containers(4).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(5).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(4).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+    containers(5).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+
+    // Make sure the created prewarmed containers are expired
+    stream.reset()
+    Thread.sleep(ttl.toMillis)
+    pool ! AdjustPrewarmedContainer
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
+
+    containers(4).expectMsg(Remove)
+    containers(5).expectMsg(Remove)
+    containers(4).send(pool, ContainerRemoved(false))
+    containers(5).send(pool, ContainerRemoved(false))
+
+    // removed previous 2 prewarmed container due to expired
+    stream.toString should include(s"removed 2 expired prewarmed container")
+
+    stream.reset()
+    // 5 code start happened(5 > maxCount)
+    pool ! run
+    pool ! run
+    pool ! run
+    pool ! run
+    pool ! run
+
+    containers(6).expectMsg(run)
+    containers(7).expectMsg(run)
+    containers(8).expectMsg(run)
+    containers(9).expectMsg(run)
+    containers(10).expectMsg(run)
+
+    pool ! AdjustPrewarmedContainer
+    // Make sure adjustPrewarmContainer run finished
+    Thread.sleep(2.seconds.toMillis)
```

Comment:
I think you can avoid most of the `Thread.sleep` usage - e.g. use eventually, or similar 
```
eventually { 
  stream.toString should include(s"currentCount: 0 prewarmed container")
}
```

Reply:
Great! updated accordingly